### PR TITLE
Rank categorical perf

### LINKED ISSUE #15498: PERF: categorical rank
xref https://github.com/pandas-dev/pandas/pull/15422#discussion_r102941126

easy enough after #15422 to rank the categories themselves rather than using expanded values; prob most relevant for ``object`` dtypes.

```
In [15]: s = Series(tm.makeCategoricalIndex(100000))

In [16]: res = Series(np.array(s.cat.rename_categories(Series(s.cat.categories).rank()))).rank()

In [17]: res2 = s.rank()

In [18]: res.equals(res2)
Out[18]: True

In [19]: %timeit Series(np.array(s.cat.rename_categories(Series(s.cat.categories).rank()))).rank()
100 loops, best of 3: 4.39 ms per loop

In [20]: %timeit s.rank()
10 loops, best of 3: 132 ms per loop
```

### PR DESCRIPTION
**Pull request by jeetjitsu**
- [x] closes #15498 
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew

### CODE REVIEW

**Comment by jreback** [comment]
also a whatsnew (in Performance) entry would be great. code lgtm.

File: pandas/tests/series/test_analytics.py [hunk 1]
```diff
@@ -1083,6 +1083,16 @@ def test_rank_categorical(self):
         res = unordered.rank()
         assert_series_equal(res, exp_unordered)
 
+        unordered1 = pd.Series(
+            [1, 2, 3, 4, 5, 6],
+        ).astype('category').cat.set_categories(
+            [1, 2, 3, 4, 5, 6],
+            ordered=False
```
Comment (jorisvandenbossche) [inline pandas/tests/series/test_analytics.py]: you can simplify this creation by doing `astype('category', categories=[1, 2, 3, 4, 5, 6], ordered=False)`

File: pandas/core/categorical.py [hunk 2]
```diff
@@ -1416,14 +1416,19 @@ def _values_for_rank(self):
...
             mask = values == -1
             if mask.any():
                 values = values.astype('float64')
                 values[mask] = np.nan
-        else:
+        elif self.categories.is_numeric():
             values = np.array(self)
+        else:
+            values = np.array(
+                self.rename_categories(Series(self.categories).rank())
```
Comment (jorisvandenbossche) [inline pandas/core/categorical.py]: Can you add a small comment on why do this here?

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good to me!

**Comment by ikilledthecat** [comment]
I have made all the minor changes but can't seem to run the asv benchmarks. Tried running them on upstream/master as well but the tests fail with the following error

TypeError: unbound methods must have non-NULL im_class

I used the following command to run asv benchmarks

 asv continuous -f 1.1 -E virtualenv upstream/master HEAD -b ^categorical

Googling for a solution hasn't helped. If you can point me towards a solution i will run the benchmarks and push my changes.

**Comment by jorisvandenbossche** [comment]
@ikilledthecat Can you push your changes anyway (regardless whether you get asv running), so we can already review (and getting the asv benchs running is not a blocker for merging this PR in this case).

Regarding the asv, I don't have experience with running it using virtualenv, only used it with conda and never saw such an error. 
As the error seems to come from asv itself, you may want to raise an issue on their tracker (https://github.com/spacetelescope/asv)

**Comment by jorisvandenbossche** [comment]
@ikilledthecat something went wrong with your rebase, as you have many other commits included now as well. Doing

```
git fetch upstream
git checkout rank_categorical_perf
git rebase upstream/master
git push -f origin rank_categorical_perf
```

should normally always do the right thing

**Comment by ikilledthecat** [comment]
@jorisvandenbossche sorry, pushed after re-mergeing upstream/master.

**Comment by jreback** [comment]
you'll want to rebase on master as travis was having some issues

**Comment by jreback** [comment]
thanks @ikilledthecat

**Pull request closed without merging**